tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
   return Status::OK();
 }
 
+// Overall, emit code for slices based on the below structure. A `slice_guard_i`
+// and a `slice_i` are generated for each slice. `slice_guard_i` computes
+// the guarding condition to decide if it should jump into `slice_i`
+// for writing to slice output or continue to next `slice_guard_{i+1}`.
+//
+// init_block:
+//   Compute values of slice input operands
+//   Br slice_guard_0
+//
+// slice_guard_0:
+//   Compute guarding_cond0
+//   CondBr guarding_cond0, slice_0, slice_guard_1
+//
+// slice_0:
+//   Write to output of slice0
+//   Br slice_guard_1
+//
+// slice_guard_1:
+//   Compute guarding_cond1
+//   CondBr guarding_cond1, slice_1, exit_block
+//
+// slice_1:
+//   Write to output of slice1
+//   Br exit_block
+//
+// exit_block:
+//
+void IrEmitterUnnested::EmitElementForInputFusibleSlices(
+    HloInstruction* unnested_hlo, const llvm_ir::IrArray::Index& index) {
+  VLOG(10) << "Emitting slice input fusion for " << unnested_hlo->ToString();
+
+  HloInstruction* slice_or_tuple = unnested_hlo->fused_expression_root();
+  auto slice_instructions = ([&]() -> absl::Span<HloInstruction* const> {
+    if (slice_or_tuple->opcode() == HloOpcode::kSlice) {
+      return absl::Span<HloInstruction* const>(&slice_or_tuple, 1);
+    }
+    DCHECK(slice_or_tuple->opcode() == HloOpcode::kTuple);
+    return slice_or_tuple->operands();
+  })();
+
+  // Emit input operand values of slices here.
+  std::vector<llvm::Value*> input_ir_values;
+  GpuElementalIrEmitter elem_emitter(hlo_module_config_, module_, &b_,
+                                     GetNestedComputer());
+  FusedIrEmitter fused_emitter(GetGeneratorForOperandIrArrays(unnested_hlo),
+                               &elem_emitter);
+  TF_CHECK_OK(unnested_hlo->fused_expression_root()->Accept(&fused_emitter));
+  for (size_t i = 0; i < slice_instructions.size(); ++i) {
+    const HloInstruction* slice = slice_instructions[i];
+    auto input_generator = fused_emitter.GetGenerator(slice->operand(0));
+    input_ir_values.push_back(input_generator(index).ValueOrDie());
+  }
+
+  // Emit for slice_instructions begins here.
+  llvm::BasicBlock* init_block = b_.GetInsertBlock();
+  llvm::BasicBlock* exit_block;
+  exit_block = init_block->splitBasicBlock(b_.GetInsertPoint(),
+                                           IrName(unnested_hlo, "merge"));
+
+  // First, generate blocks and guard_blocks for each slice. Will populate
+  // them below.
+  std::vector<llvm::BasicBlock*> emitted_guard_blocks;
+  std::vector<llvm::BasicBlock*> emitted_slice_blocks;
```

Comment:
Optional: it's customary to do `.reserve(slice_instructions.size())` in such cases.

Reply:
Added.